orttraining/orttraining/training_ops/cuda/tensor/concat.cc

```diff
@@ -50,26 +50,59 @@ Status ConcatTraining::ComputeInternal(OpKernelContext* ctx) const {
     concat_sizes_range[i] += concat_sizes_range[i - 1];
   }
 
-  CudaAsyncBuffer<int64_t> concat_sizes_gpu(this, concat_sizes);
-  CudaAsyncBuffer<int64_t> axis_dimension_input_output_mapping_gpu(this, axis_dimension_input_output_mapping);
-  CudaAsyncBuffer<int64_t> concat_sizes_range_gpu(this, concat_sizes_range);
-  concat_sizes_gpu.CopyToGpu();
-  axis_dimension_input_output_mapping_gpu.CopyToGpu();
-  concat_sizes_range_gpu.CopyToGpu();
-  input_ptr.CopyToGpu();
+  auto element_bytes = p.output_tensor->DataType()->Size();
   int block_size_inside_axis_dim = static_cast<int>(p.output_axis_pitch / p.output_tensor->Shape()[p.axis]);
   int block_size_including_axis_dim = static_cast<int>(p.output_axis_pitch);
-  auto element_bytes = p.output_tensor->DataType()->Size();
-  ORT_RETURN_IF_ERROR(ConcatImpl(Stream(),
+  if (std::all_of(concat_sizes.begin(), concat_sizes.end(), [&] (int64_t i) {return i == concat_sizes[0];})) {
+    if (input_count <= 32) {
+      // pass by value to avoid host-to-device copy on same stream
+      TArray<const void*, 32> input_table(input_count);
+      for (int i = 0; i < input_count; ++i) {
+	input_table[i] = input_ptr_cpuspan[i];
+      }
+      ORT_RETURN_IF_ERROR(ConcatSameConcatDimImpl(Stream(),
+                                 element_bytes,
+                                 block_size_including_axis_dim,
+                                 block_size_inside_axis_dim,
+                                 concat_sizes[0],
+                                 p.output_tensor->MutableDataRaw(),
+                                 input_table,
+                                 p.output_num_elements));
+ 
+    } else {
+      // too many inputs, so copy sizes to device memory
+      input_ptr.CopyToGpu();
```

Comment:
For ProphetNet, did you see the perf diff between using TArray vs non-TArray? Actually, the kernel with TArray used here would be slower than the kernel w/o TArray since GPU local memory will be used for TArray for this scenario.

Reply:
I'm not sure local memory is used much for these operators? To be safe, I just did a run with 500 steps for ProphetNet with ORT build off master. Using TArray to pass by value got 1.09 steps/sec vs. copying by buffer got 1.00 steps/sec. I tested it a couple times. So about +9% with the TArray.

Reply:
On CUDA V100 the TArray only provides a marginal benefit - about 0.5%. And the overall gain is 1.5%.

> > This is nice! @mindest and me also thought about part of the optimization but we fail to see big gains. Just curious, opt the same dim cases I assume give the most of the 15% gains?
> 
> I believe cutting down on extra kernel launches is the larger part of the gain. What was your test case and finding? There should be some gain on CUDA as well, but it won't be as large since ROCm suffers a greater kernel dispatch overhead than CUDA.



> > This is nice! @mindest and me also thought about part of the optimization but we fail to see big gains. Just curious, opt the same dim cases I assume give the most of the 15% gains?
> 
> I believe cutting down on extra kernel launches is the larger part of the gain. What was your test case and finding? There should be some gain on CUDA as well, but it won't be as large since ROCm suffers a greater kernel dispatch overhead than CUDA.

Yes, the gain for CUDA V100 is more modest - only 1.5%. Much of the improvement here is mitigating kernel dispatch effects on ROCm stack.